xenomint/state.go

```diff
@@ -0,0 +1,502 @@
+/*
+ * Copyright 2018 The CovenantSQL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenomint
+
+import (
+	"database/sql"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/CovenantSQL/CovenantSQL/types"
+	xi "github.com/CovenantSQL/CovenantSQL/xenomint/interfaces"
+	"github.com/pkg/errors"
+)
+
+type State struct {
+	sync.RWMutex
+	strg xi.Storage
+	pool *pool
+
+	// TODO(leventeliu): Reload savepoint from last block on chain initialization, and rollback
+	// any ongoing transaction on exit.
+	//
+	// unc is the uncommitted transaction.
+	unc     *sql.Tx
+	origin  uint64 // origin is the original savepoint of the current transaction
+	cmpoint uint64 // cmpoint is the last commit point of the current transaction
+	current uint64 // current is the current savepoint of the current transaction
+}
+
+func NewState(strg xi.Storage) (s *State, err error) {
+	var t = &State{
+		strg: strg,
+		pool: newPool(),
+	}
+	if t.unc, err = t.strg.Writer().Begin(); err != nil {
+		return
+	}
+	t.setSavepoint()
+	s = t
+	return
+}
+
+func (s *State) incSeq() {
+	s.current++
+}
+
+func (s *State) setNextTxID() {
+	var val = (s.current & uint64(0xffffffff00000000)) + uint64(1)<<32
+	s.origin = val
+	s.cmpoint = val
+	s.current = val
+}
+
+func (s *State) setCommitPoint() {
+	var val = (s.current & uint64(0xffffffff00000000)) + uint64(1)<<32
+	s.cmpoint = val
+	s.current = val
+}
+
+func (s *State) rollbackID(id uint64) {
+	s.current = id
+}
+
+func (s *State) InitTx(id int32) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.InitTx should have comment or be unexported